lib/transports/ws2.js

```diff
@@ -0,0 +1,832 @@
+      this._eventCallbacks.push(key, (err, res) => {
+        if (err) {
+          return reject(err)
+        }
+
+        resolve(res)
+      })
+    })
+  }
+
+  send (msg) {
+    if (!this._ws) throw new Error('ws not open')
+
+    debug('sending %j', msg)
+    this._ws.send(JSON.stringify(msg))
+  }
+
+  enableSequencing () {
+    this.send({
+      event: 'conf',
+      flags: 65536
+    })
+  }
+
+  onServerRestart (cb) {
+    this.onInfoMessage(WSv2.info.SERVER_RESTART, cb)
+  }
+
+  onMaintenanceStart (cb) {
+    this.onInfoMessage(WSv2.info.MAINTENANCE_START, cb)
+  }
+
+  onMaintenanceEnd (cb) {
+    this.onInfoMessage(WSv2.info.MAINTENANCE_END, cb)
+  }
+
+  /**
+   * @param {string} channel
+   * @param {Object} payload - optional
+   */
+  subscribe (channel, payload) {
+    this.send(Object.assign({
+      event: 'subscribe',
+      channel
+    }, payload))
+  }
+
+  /**
+   * @param {string} symbol
+   */
+  subscribeTicker (symbol) {
+    return this.managedSubscribe('ticker', symbol, { symbol })
+  }
+
+  /**
+   * @param {string} symbol
+   */
+  subscribeTrades (symbol) {
+    return this.managedSubscribe('trades', symbol, { symbol })
+  }
+
+  /**
+   * @param {string} symbol
+   * @param {string} prec - P0, P1, P2, or P3 (default P0)
+   * @param {string} len - 25 or 100 (default 25)
+   */
+  subscribeOrderBook (symbol, prec = 'P0', len = '25') {
+    return this.managedSubscribe('book', symbol, { len, prec })
+  }
+
+  /**
+   * @param {string} symbol
+   * @param {string} frame - time frame
+   */
+  subscribeCandles (symbol, frame) {
+    const key = `trade:${frame}:${symbol}`
+
+    return this.managedSubscribe('candles', key, { key })
+  }
+
+  /**
+   * @param {number} chanId
+   */
+  unsubscribe (chanId) {
+    this.send({
+      event: 'unsubscribe',
+      chanId: +chanId
+    })
+  }
+
+  /**
+   * @param {string} symbol
+   */
+  unsubscribeTicker (symbol) {
+    return this.managedUnsubscribe('ticker', symbol)
+  }
+
+  /**
+   * @param {string} symbol
+   */
+  unsubscribeTrades (symbol) {
+    return this.managedUnsubscribe('trades', symbol)
+  }
+
+  /**
+   * @param {string} symbol
+   * @param {string} prec - P0, P1, P2, or P3 (default P0)
+   * @param {string} len - 25 or 100 (default 25)
+   */
+  unsubscribeOrderBook (symbol, prec = 'P0', len = '25') {
+    return this.managedUnsubscribe('book', symbol)
+  }
+
+  /**
+   * @param {string} symbol
+   * @param {string} frame - time frame
+   */
+  unsubscribeCandles (symbol, frame) {
+    return this.managedUnsubscribe('candles', `trade:${frame}:${symbol}`)
+  }
+
+  /**
+   * @param {string} cbGID
+   */
+  removeListeners (cbGID) {
+    delete this._listeners[cbGID]
+  }
+
+  // Throttle to 8 op/s
+  requestCalc () {
+    throw new Error('unimplemented') // TODO
+  }
+
+  /**
+   * @param {Object|Array} order
+   * @return {Promise} p
+   */
+  submitOrder (order) {
+    if (!this._isAuthenticated) throw new Error('not authenticated')
```

Comment:
emit error instead?

Reply:
oh wait, we return a promise later. then retun a promise and reject it?

maybe like this?

```js
if (!this._isAuthenticated) return Promise.reject()

```